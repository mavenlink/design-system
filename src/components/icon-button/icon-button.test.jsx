import React, {
  createRef,
} from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import user from '@testing-library/user-event';
import IconButton from './icon-button.jsx';

describe('<IconButton />', () => {
  const requiredProps = {
    icon: {
      id: 'test-id',
      viewBox: '1 2 3 4',
    },
    label: 'Test label',
    onPress: () => {},
  };

  it('has defaults', () => {
    const ref = createRef();
    render(<IconButton {...requiredProps} ref={ref} />);
    expect(document.body).toMatchSnapshot();
    expect(document.body).toBe(document.activeElement);
    expect(ref.current).toMatchSnapshot();
  });

  describe('active API', () => {
    it('can be set', () => {
      render(<IconButton {...requiredProps} active />);
      expect(screen.getByRole('button')).toHaveAttribute('tabindex', '0');
    });

    it('can be unset', () => {
      render(<IconButton {...requiredProps} active={false} />);
      expect(screen.getByRole('button')).toHaveAttribute('tabindex', '-1');
    });
  });

  describe('className API', () => {
    it('can be set', () => {
      render(<IconButton {...requiredProps} className="unique-class-name" />);
      expect(screen.getByRole('button', { name: 'Test label' })).toHaveClass('unique-class-name');
    });
  });

  describe('icon API', () => {
    it('sets the xlink:href', () => {
      render(<IconButton {...requiredProps} icon={{ ...requiredProps.icon, id: 'unique-id' }} />);
      expect(screen.getByRole('button', { name: 'Test label' }).children[1]).toHaveAttribute('xlink:href', '#unique-id');
    });

    it('sets the height', () => {
      render(<IconButton {...requiredProps} icon={{ ...requiredProps.icon, viewBox: '1 2 3 41' }} />);
      expect(screen.getByRole('button', { name: 'Test label' })).toHaveAttribute('height', '41');
    });

    it('sets the width', () => {
      render(<IconButton {...requiredProps} icon={{ ...requiredProps.icon, viewBox: '1 2 42 4' }} />);
      expect(screen.getByRole('button', { name: 'Test label' })).toHaveAttribute('width', '42');
    });
  });

  describe('id API', () => {
    it('can be set', () => {
      const ref = createRef();

      render((
        <React.Fragment>
          <input aria-labelledby="unique-id" ref={ref} />
          <IconButton {...requiredProps} label="unique label" id="unique-id" />
        </React.Fragment>
      ));

      expect(screen.getByLabelText('unique label')).toBe(ref.current);
    });
  });

  describe('label API', () => {
    it('can be set', () => {
      render(<IconButton {...requiredProps} label="Unique label" />);
      expect(screen.getByRole('button', { name: 'Unique label' })).toBeInTheDocument();
    });
  });

  describe('labelledBy API', () => {
    it('can be set', () => {
      render((
        <React.Fragment>
          <label id="label-id">Unique label</label>
          <IconButton {...requiredProps} labelledBy="label-id" />
        </React.Fragment>
      ));
      expect(screen.getByLabelText('Unique label')).toBeInTheDocument();
    });
  });

  describe('onPress API', () => {
    it('is called on click', () => {
      const onPressSpy = jest.fn(event => event.persist());
      render(<IconButton {...requiredProps} onPress={onPressSpy} role="button" />);
      user.click(screen.getByRole('button'));
      expect(onPressSpy).toBeCalledWith(expect.objectContaining({ target: expect.anything() }));
    });

    it('is called on enter keypress', () => {
      const onPressSpy = jest.fn(event => event.persist());
      render(<IconButton {...requiredProps} onPress={onPressSpy} role="button" />);
      user.tab();
      user.keyboard('{Enter}');
      expect(onPressSpy).toBeCalledWith(expect.objectContaining({
        defaultPrevented: true,
        target: expect.anything(),
      }));
    });

    it('is called on space keypress', () => {
      const onPressSpy = jest.fn(event => event.persist());
      render(<IconButton {...requiredProps} onPress={onPressSpy} role="button" />);
      user.tab();
      user.keyboard(' ');
      expect(onPressSpy).toBeCalledWith(expect.objectContaining({
        defaultPrevented: true,
        target: expect.anything(),
      }));
    });
  });
});
