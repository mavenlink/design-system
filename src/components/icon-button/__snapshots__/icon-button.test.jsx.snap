// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> has defaults 1`] = `
<body>
  <div>
    <svg
      aria-disabled="false"
      class="button"
      height="4"
      role="button"
      tabindex="0"
      width="3"
    >
      <title>
        Test label
      </title>
      <use
        xlink:href="#test-id"
      />
    </svg>
  </div>
</body>
`;

exports[`<IconButton /> has defaults 2`] = `
<svg
  aria-disabled="false"
  class="button"
  height="4"
  role="button"
  tabindex="0"
  width="3"
>
  <title>
    Test label
  </title>
  <use
    xlink:href="#test-id"
  />
</svg>
`;
